GEP-1911 - Backend Protocol Selection (#1979)
kubernetes-sigs/gateway-api: geps/gep-1911.md

### Default Protocols

If a Service `appProtocol` isn't specified an implementation MAY infer the backend protocol through its own means. Implementations MAY infer the protocol from the `Route` type referring to the backend Service.

### Multiple Protocols on the same port

Only the Kubernetes `Service` `protocol` field supports multiple protocols on the same port. See the details in [KEP-1435][kep-1435].

Implementations MUST recognize support for `TCP` and `UDP` multiplexing on the same port.

Currently Kubernetes `Service` API only allows different `appProtocol` values for the same port when `protocol` fields differs. At this time there seems to be interest in changing `appProtocol` to be a list in order to faciliate this use-case.

### Unsupported/Incompatible Protocols

If a Route attached to a Gateway is not able to send traffic to the backend using the specified protocol then the backend is considered invalid. Implementations MUST set `ResolvedRefs` condition to `False` with the Reason `UnsupportedProtocol`.

> Implementations MUST supported the following combinations below:

Reviewer: I don't think that we can require implementations to support these combinations. We can recommend that they do, or even just state these combinations have well-defined interactions with Gateway API and should be supported wherever possible.

Author: Should I change the entire GEP to be `RECOMMENDED` ? line 37 has a `MUST`

Reviewer 2: this section has some language that might help, talking about Listener protocols so a similar idea: https://github.com/kubernetes-sigs/gateway-api/blob/ea382e7a989d2849ce1a76d47a60bd5f5ea754e5/apis/v1beta1/gateway_types.go#L260-L266

(though the line referencing a table in the linked godoc might need to be removed now)

I think in other places `MAY` is used in similar contexts

If we anticipate protocol/kind combos in the tables to be tested in conformance then could add a mention of that? For standardization these combinations will be verified, but as extended conformance?

Author: >  but as extended conformance so not a MUST?

OK sounds good

Author: Circling back - I opt'd to leave the `MUST` requirement that implementations look at the `appProtocol` field to infer backend protocol.

And I changed the specific `support {x} protocol with {y} route` an optional (MAY) requirement

ServicePort Protocol | ServicePort AppProtocol | Route Type | Supported
-|-|-|-
`TCP` | `kubernetes.io/h2c` | `GRPCRoute` | Yes [1]
`TCP` | `kubernetes.io/h2c` | `HTTPRoute` | Yes
`TCP` | `kubernetes.io/ws`  | `HTTPRoute` | Yes

> `TCP` | `kubernetes.io/wss` | `TLSRoute`  | Yes

Reviewer: Does the absence of this AppProtocol fields mean that we shouldn't enable any websocket upgrades? What about implementations that already have this upgrade enabled by default in their underlying dataplane?

Author: > Does the absence of this AppProtocol fields mean that we shouldn't enable any websocket upgrades? 

No - this

> What about implementations that already have this upgrade enabled by default in their underlying dataplane?

No change - ideally users wouldn't need to set these `appProtocol` values and things just work :) 

I'll add some language to the `Default Protocol` section

Author: Added

> Absence of the `appProtocol` field does not imply the implementation should disable any features (eg. websocket upgrades).

1. GRPC works over h2c - so a GRPCRoute should be able to connect to an h2c backend

Implementations MAY support the following combinations below:

ServicePort Protocol | ServicePort AppProtocol | Route Type | Supported
-|-|-|-
ServicePort Protocol | ServicePort AppProtocol | Route Type | Supported
`TCP`  | `kubernetes.io/h2c` | `TLSRoute`  | Conditional [1]

> `TCP`  | `kubernetes.io/ws`  | `TLSRoute`  | Conditional [1]

Reviewer: These feel a little wonky to me. Now that we're moving to "allow-list" only, I think I'd be in favor of omitting these unless we have a clear need for them. For example, we've said that [HTTPRoute can't be used with a TLS listener protocol](https://gateway-api.sigs.k8s.io/guides/tls/#clientserver-and-tls). With that context, it feels weird to say that you can use h2c or ws with TLSRoute. (I recognize that we haven't actually directly said anything that would prevent the use of HTTP from a Gateway to a Backend with TLS as the listener protocol, just HTTPRoute, but it all feels awfully unsettled and underspecified right now and I'd hate to add to the potential confusion here).

## Open Questions

1. TLSRoute & UDP protocol

TLS over UDP seems to be a thing via QUIC/HTTP3 [ref](https://www.smashingmagazine.com/2021/08/http3-core-concepts-part1/).
Likewise there's also [DTLS](https://en.wikipedia.org/wiki/Datagram_Transport_Layer_Security). But it's unclear if Gateway's TLSRoute
applies to an underlying UDP protocol.

2. Websockets & HTTP/2/3

Should we upstream new constants for websocket over [HTTP/2](https://www.rfc-editor.org/rfc/rfc8441.html) & [HTTP/3](https://www.rfc-editor.org/rfc/rfc9220.html) ? HTTP/3 makes things more complicated since its supports UDP as the underlying protocol.

## Alternatives

### Single Meta-resource 

The first pass of this GEP proposed a new meta-resource [GEP-713](geps/gep-713) called `BackendProtocol`.

This allows end-users to specify a list of ports and a list of corresponding protocols that that single
port supports.

This was dropped in favour of supporting Kubernetes Standard Application Protocols.

### Multiple Protocol Meta-resources

Rather than bundle protocol details into a single resource an alternative would be to create distinct meta resources.
ie. `HTTP2Backend`, `GPRCBackend`, `WebsocketBackend`.

> The advantages of this approach are:
> - Easy to introduce new protocols
> - Definitions/types would be simpler
> 
> The disadvantages of this approach are:
> - N resources for N protocols need to be created to describe a single backend
> - No easy mechanic to specify priority of protocols

Reviewer: New lines needed for list formatting.
```suggestion
The advantages of this approach are:

- Easy to introduce new protocols
- Definitions/types would be simpler

The disadvantages of this approach are:

- N resources for N protocols need to be created to describe a single backend
- No easy mechanic to specify priority of protocols
```

  //  ports: 
  //   - port: 80
  //     http2: 
  //       encrypted: false # enable h2c
  //   - port: 80
  //     http1: {}          # We also listen to http1 on port 80
  //
  Ports []BackendPort `json:"ports,omitempty"`
}

type BackendPort struct {
  // Number or name of the port who's protocol is being defined
  Port intstr.IntOrString

  PortProtocol `json:",inline"`
}

type PortProtocol struct {
  HTTP2     *HTTP2PortProtocol    `json:"http2"`
  HTTP1     *HTTP1PortProtocol    `json:"http1"`
  GRPC      *GRPCPortProtocol     `json:"grpc"`
  WebSocket *WebsocketPortProtocol`json:"websocket"`
}

type HTTP2PortProtocol struct {
  Cleartext bool // h2c when set to true

  // Future knobs can go here
}

type HTTPPortProtocol struct {
  // Future knobs can go here
}
type GRPCPortProtocol struct {
  // Future knobs can go here
}
type WebsocketPortProtocol struct {
  // Future knobs can go here
}
```

### YAML API

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha1
kind: BackendProtocol
metadata:
  name: some-service
  namespace: some-namespace
spec:
  ports:
  - port: 80
    http2: 
      encrypted: false # enable h2c
  - port: 80

>     http1: {}          # We also listen to http1 on port 80

Author: Probably worth calling out you could pack multiple protocol entries in a single `BackendPort` struct - but you would lose priority between protocols (unless we added another mechanic)

```yaml
- port: 80
  http2: {}
  http1: {}
```

If we choose to limit `BackendPort` struct to a single protocol I would need to spell that out in the GEP

# GEP-1911: Backend Protocol Selection

* Issue: [#1911](https://github.com/kubernetes-sigs/gateway-api/issues/1911)
* Status: Provisional

(See status definitions [here](overview.md#status).)

## TLDR

> Not all implementation support automatic protocol selection. Even in some cases protocols are disabled without an explicit opt-in (ie. websockets). Thus application developers need the ability specify the protocol(s) that their application supports.
> 

Reviewer 3: ```suggestion
Not all implementations support automatic protocol selection. Even in some cases protocols are disabled without an explicit opt-in (e.g. websockets). Thus application developers need the ability to specify the protocol(s) that their application supports.

```

## Goals

- Support protocols that can have a Gateway '*Route' resource as a frontend
- Consistent way of declaring backend protocols using a meta-resource
- Support backends with multiple protocols on the same port (ie. h2c/http1)
- Extensible - additional protocol specific fields should be easy to add in the future

## Non-Goals

- Backend TLS (covered in [GEP-1897](https://github.com/kubernetes-sigs/gateway-api/issues/1897)

## Introduction

## API

> We propose adding a new meta-resource [GEP-713](geps/gep-713.md) called `BackendProtocol`.

Reviewer 3: Is #713 the right GEP?

Author: This is open for discussion. Technically since I'm using a `targetRef` it could be _anything_. But if we want to scope this to just `Service` then we could simplify this.

As is,  _we_ could target a multi cluster ServiceImport - but I'm not to familiar with that. cc @robscott

Author: Re-wrote the GEP so it's not a meta-resource anymore

  // Number or name of the port who's protocol is being defined
  Port intstr.IntOrString

  PortProtocol `json:",inline"`
}

type PortProtocol struct {
  HTTP2     *HTTP2PortProtocol    `json:"http2"`
  HTTP1     *HTTP1PortProtocol    `json:"http1"`
  GRPC      *GRPCPortProtocol     `json:"grpc"`
  WebSocket *WebsocketPortProtocol`json:"websocket"`
}

type HTTP2PortProtocol struct {
  Cleartext bool // h2c when set to true

  // Future knobs can go here
}

type HTTPPortProtocol struct {
  // Future knobs can go here
}
type GRPCPortProtocol struct {
  // Future knobs can go here
}
type WebsocketPortProtocol struct {
  // Future knobs can go here
}
```

### YAML API

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha1
kind: BackendProtocol
metadata:
  name: some-service
  namespace: some-namespace
spec:
  ports:
  - port: 80
    http2: 
      encrypted: false # enable h2c
  - port: 80
    http1: {}          # We also listen to http1 on port 80
  - port: 9090
    grpc: {}
  - port: 8080
    websocket: {}      # enable websocket support on this port
  - port: 7070
    http1: {}
  targetRef:
    group: "" # corev1
    kind: Service
    name: some-service
```

### Default Protocols

> If a backend protocol isn't specified an implementation MAY infer the backend protocol through it's own means.

Reviewer 3: ```suggestion
If a backend protocol isn't specified an implementation MAY infer the backend protocol through its own means.
```

### Incompatible Protocols

If a Route attached to a Gateway is not able to support sending traffic to the backend using a specific protocol 
then it must set it's condition to `Accepted` `False`.

## Alternatives

### Meta-resource for each protocol

Rather than bundle protocol details into a single resource an alternative would be to create distinct meta resources.
ie. `HTTP2Backend`, `GPRCBackend`, `WebsocketBackend`.

> Although this has the advantage of seperate it doesn't allow

Reviewer 3: Is there more to this? :)

Author: whoops

Author: flushed this out a bit

### Adding properties on Gateway Route Objects

From [GEP-1282](geps/gep-1282.md#tldr):
> some types of configuration requested by users are more about defining functionality that describes capabilities of the backend more than the route you take to get to the backend.

Backend protocol is specifying capabilities. This configuration is less about routing.

### Extending Kubernetes Service

This is considered untenable due to the 'the turnaround time for those changes can be years.' ([ref-1282](geps/gep-1282.md#non-goals):

### AppProtocol

From [GEP-1282](geps/gep-1282.md#non-goals):
> v1.Service’s appProtocol field is not fit for purpose, because it is defined as accepting values either from the IANA Service Name registry, or domain-prefixed values and we need more flexibility than that. 

> See further discussion [here](https://github.com/kubernetes-sigs/gateway-api/discussions/1244#discussioncomment-3151197)

Reviewer 4: I am not sure this is true since https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3726-standard-application-protocols#summary?

There was talk of allowing multiple appProtocols as well - I don't think it landed but I don't think its in-plausible

Reviewer: Yep I think multiple app protocols is inevitable - help wanted though. @LiorLieberman has been helping move AppProtocols forward upstream.

Reviewer 5: this would be needed for all backend types including ServiceImport
https://github.com/kubernetes-sigs/mcs-api/blob/11eea5e71266160cc900f319b3f080c24dbea4f1/pkg/apis/v1alpha1/serviceimport.go#L99

Author: > There was talk of allowing multiple appProtocols as well - I don't think it landed but I don't think its in-plausible

Is that this https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1435-mixed-protocol-lb ? It's GA in K8s 1.26. Do we know why the KEP is scoped to type=LoadBalancer?

> I am not sure this is true since https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3726-standard-application-protocols#summary?

The examples are confusing - it suggests `HTTP2` has two different meanings (one is http2 over tls in GKE and the other is h2c with Istio). Not a fan of that.

I think we need to decided do we want the ability to customize the backend in any way - https://github.com/kubernetes-sigs/gateway-api/pull/1979#issuecomment-1526585087

Reviewer: >The examples are confusing - it suggests HTTP2 has two different meanings (one is http2 over tls in GKE and the other is h2c with Istio). Not a fan of that.

To clarify, that was trying to note that the same values were being interpreted different ways by different implementations. It was not an example of something to do, more something this KEP was trying to solve.

Author: I shouldn't skim KEPs during dinner ;)

Author: One observation I have about appProtocol (or `lists of protocols` in general) is how can a backend _disable_ a protocol.

ie. disable websockets & http

Author: Added some comments in the GEP from this discussion

Following up on[KEP-1435 - Mixed Protocol](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1435-mixed-protocol-lb) - it only applies to the `protocol` field. Thus trying to specify a K8s Service with multiple `appProtocol` to a single port won't work

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
  labels:
    app.kubernetes.io/name: proxy
spec:
  containers:
  - name: nginx
    image: nginx:stable
    ports:
      - containerPort: 80
        name: http-web-svc
---
apiVersion: v1
kind: Service
metadata:
  name: nginx-service
spec:
  selector:
    app.kubernetes.io/name: proxy
  ports:
  - name: http
    protocol: TCP
    appProtocol: http
    port: 80
    targetPort: http-web-svc
  - name: h2c
    appProtocol: h2c
    protocol: TCP
    port: 80
    targetPort: http-web-svc
```

Author: Switched the GEP to consider `appProtocol` resolving - please review

An option would be to have a specific vendor prefix but unfortunately this limits the protocol to one per port.

### Unstructured Data/Special Values

Unstructured data refers to using labels and annotations.

From [GEP-1282](geps/gep-1282.md#non-goals):

> > these are very sticky and hard to get rid of once you start using them.

Reviewer 3: Annotations are also not self-documenting, there is no `kubectl explain` for annotations.

Author: added this to the section

  // When multiple protocols reference the same port the 
  // entries earlier in the list should have higher priority
  //
  // ie. attempt to use h2c otherwise fallback on http1
  //  ports: 
  //   - port: 80
  //     http2: 
  //       encrypted: false # enable h2c
  //   - port: 80
  //     http1: {}          # We also listen to http1 on port 80
  //
  Ports []BackendPort `json:"ports,omitempty"`
}

type BackendPort struct {
  // Number or name of the port who's protocol is being defined
  Port intstr.IntOrString

  PortProtocol `json:",inline"`
}

type PortProtocol struct {
  HTTP2     *HTTP2PortProtocol    `json:"http2"`
  HTTP1     *HTTP1PortProtocol    `json:"http1"`
  GRPC      *GRPCPortProtocol     `json:"grpc"`
  WebSocket *WebsocketPortProtocol`json:"websocket"`
}

type HTTP2PortProtocol struct {
  Cleartext bool // h2c when set to true

  // Future knobs can go here
}

type HTTPPortProtocol struct {
  // Future knobs can go here
}
type GRPCPortProtocol struct {
  // Future knobs can go here
}
type WebsocketPortProtocol struct {
  // Future knobs can go here
}
```

### YAML API

```yaml
apiVersion: gateway.networking.k8s.io/v1alpha1

> kind: BackendProtocol

Reviewer 6: The service is k8s core/v1 has already include ports and protocol,, why not use  it

### New Protocols & Reserved Prefix

To add support for a new protocol it should first become a Kubernetes Standard Application Protocol by updating the [KEP-3726][kep-3726]. If the suggested protocol is not suited to have a `kubernetes.io/*` prefix then the Gateway API MAY support the new protocol using it's own prefix `gateway.networking.k8s.io/*`.  Please make a PR to this GEP.

For example we may want to add a sentinel `appProtocol` value that prevents Gateway implementations from discovering the protocol of the application. Instead they should just refer to the Service's `protocol` field. Such a constant was rejected upstream (https://github.com/kubernetes/enhancements/pull/4106) but as an example it could be defined in this GEP as `gateway.networking.k8s.io/no-sniff`. 

### Default Protocols

If a Service `appProtocol` isn't specified an implementation MAY infer the backend protocol through its own means. Implementations MAY infer the protocol from the `Route` type referring to the backend Service.

### Multiple Protocols on the same port

Only the Kubernetes `Service` `protocol` field supports multiple protocols on the same port. See the details in [KEP-1435][kep-1435].

> Implementations MUST recognize support for `TCP` and `UDP` multiplexing on the same port.

Author: > What does this mean? 

Gateways support K8s Service backends with different `protocol` values on the same port. Thus in theory I could point a TCPRoute and UDPRoute the referenced backend and they should be Accepted & Programmed.

>  In upstream Kubernetes the only requirement is to surface an error message if you can't support more than one protocol on the same port.

Are you suggesting we do the same?

> Implementations MUST recognize support for `TCP` and `UDP` multiplexing on the same port.

Reviewer: >Are you suggesting we do the same?

I'm suggesting that we shouldn't have stricter requirements than Kubernetes upstream. I don't think we can require that every implementation of this API support multiplexing, so I think that would mean that implementations would have to surface some kind of error if they can't support it. Maybe we can standardize on what that is at least?

> Currently Kubernetes `Service` API does not allow different `appProtocol` values for the same port. At this time there seems to be interest in changing `appProtocol` to be a list in order to faciliate this use-case.

Author: Good catch - I can note the exception when the `protocol` is different

> Open Question: Can SCTP be multiplexed with TCP and UDP on the same port?

Author: i'll drop the question since we inverted the table to be an allow list only

### Unsupported/Incompatible Protocols

If a Route attached to a Gateway is not able to send traffic to the backend using the specified protocol then the backend is considered invalid. Implementations MUST set `ResolvedRefs` condition to `False` with the Reason `UnsupportedProtocol`.

> It is RECOMMENDED that implementaions follow the supported combinations below:

Author: I changed the table to be MUST for obvious things and MAY for some conditional things

1. Open Question: Should we upstream new constants for websocket over [HTTP/2](https://www.rfc-editor.org/rfc/rfc8441.html)
    & [HTTP/3](https://www.rfc-editor.org/rfc/rfc9220.html) ?

ServicePort Protocol | ServicePort AppProtocol | Route Type | Supported
-|-|-|-
`TCP`  |`kubernetes.io/h2c` | `HTTPRoute` | Yes
`UDP`  |`kubernetes.io/h2c` | `HTTPRoute` | No
`TCP`  |`kubernetes.io/h2c` | `GRPCRoute` | Yes [2]
`UDP`  |`kubernetes.io/h2c` | `GRPCRoute` | No
`UDP`  |`kubernetes.io/h2c` | `UDPRoute`  | No
`TCP`  |`kubernetes.io/h2c` | `TCPRoute`  | TBD [4]
`TCP`  |`kubernetes.io/h2c` | `TLSRoute`  | Conditional [5]
`TCP`  |`kubernetes.io/ws`  | `HTTPRoute` | Yes
`UDP`  |`kubernetes.io/ws`  | `HTTPRoute` | TBD [3]
`TCP`  |`kubernetes.io/ws`  | `GRPCRoute` | No
`UDP`  |`kubernetes.io/ws`  | `GRPCRoute` | No
`TCP`  |`kubernetes.io/ws`  | `TCPRoute`  | TBD [4]
`TCP`  |`kubernetes.io/ws`  | `TLSRoute`  | Conditional [5]
`TCP`  |`kubernetes.io/wss` | `HTTPRoute` | Conditional [1]
`UDP`  |`kubernetes.io/wss` | `HTTPRoute` | TBD [3]
`TCP`  |`kubernetes.io/wss` | `GRPCRoute` | No
`UDP`  |`kubernetes.io/wss` | `GRPCRoute` | No
`TCP`  |`kubernetes.io/wss` | `TCPRoute`  | TBD [4]

> `TCP`  |`kubernetes.io/wss` | `TLSRoute`  | Yes

Reviewer: This is pretty complex, I think it might be simpler to just have a smaller list of the combinations that we think should always be valid (`Yes`) in this table, and a separate list of the combinations that we think _may_ be valid (`Conditional`) in this table. I think that would mean we could also replace the tables above this one as well with just focusing on combinations that may be valid.

Author: I like this approach - this leaves other situations 'undefined' for now

> 1. Only if there is a corresponding `TLSConnectionPolicy` - see [GEP-1897](geps/gep-1897)

Reviewer 7: This name got changed - it's now `BackendTLSPolicy` again...

> 1. Only if there is a corresponding `TLSConnectionPolicy` - see [GEP-1897](TODO)

Reviewer 7: This doesn't make sense to me. If the backend protocol is `kubernetes.io/h2c` or `kubernetes.io/ws`, that says that the service is expecting unencrypted communication, which would rule out what's now called a `BackendTLSPolicy` (since that specifies that the service is expecting encrypted communication).

I'm still not a fan of allowing TLSRoutes with `tls.mode: terminate`, but I've lost that one, so _only because_ we allow that, I think that unencrypted protocols with TLSRoute makes sense.

So, I guess my requested change here is that the condition be updated to:

Only if the TLSRoute is attached to a Listener that has `tls.mode: terminate` set.

Author: updated please take a look

> 2. GRPC works over h2c

Reviewer 7: Is it worth calling out that doing unencrypted gRPC over an external connection like a Gateway is pretty weird though?

Author: > Is it worth calling out that doing unencrypted gRPC over an external connection like a Gateway is pretty weird though?

I just wanted to highlight GRPC can happen over clear text. I updated the clause to be more specific that the clear text connection is from the gateway to the backend.

# GEP-1911: Backend Protocol Selection

* Issue: [#1911](https://github.com/kubernetes-sigs/gateway-api/issues/1911)
* Status: Provisional

(See status definitions [here](overview.md#status).)

## TLDR

> Not all implementations support automatic protocol selection. Even in some cases protocols are disabled without an explicit opt-in (eg. websockets). Thus application developers need the ability to specify the protocol(s) that their application supports.

Reviewer 8: > Even in some cases protocols are disabled without an explicit opt-in (eg. websockets)

What do you mean here ? can you give an example?

Author: Contour by default when using their HTTPProxy resource doesn't support websockets out of the box. You need to set a bool attribute to true - details here: https://projectcontour.io/docs/1.25/config/websockets/

Author: Mentioned nginx and contour in this tl;dr where the websocket protocols were off by default

## Goals

- Support protocols that can have a Gateway '*Route' resource as a frontend
- Consistent way of declaring backend protocols using a meta-resource

> - Support backends with multiple protocols on the same port (ie. h2c/http1)

Author: > Can you elaborate on your specific case ? do you need a backend that supports multiple protocols on the same port?

Yeah this is fairly common. For example I'm aware of users running TensorFlow Serving on CloudRun. It exposes HTTP and gRPC protocols for model prediction and it's being run on a single port that's multiplexing between them.

> - Support backends with multiple protocols on the same port (ie. h2c/http1)

Author: Also K8s supports multiple Service `protocol` on the same port.

I added a section about this - please re-review.

> - Support backends with multiple protocols on the same port (ie. h2c/http1)

Reviewer 8: > Yeah this is fairly common. For example I'm aware of users running TensorFlow Serving on CloudRun. It exposes HTTP and gRPC protocols for model prediction and it's being run on a single port that's multiplexing between them.

I read a little bit about TensorFlow Service, it looks like it is two different ports. 8500 for GRPC and 8501 for HTTP.

The reason I am bringing this up is that I am trying to collect real uses cases for converting `AppProtocol` to a list.

> - Support backends with multiple protocols on the same port (ie. h2c/http1)

Reviewer 8: Also, it is still a goal on this GEP, not sure I found how this GEP suggests to support it?
few service protocols?

> - Support backends with multiple protocols on the same port (ie. h2c/http1)

Author: > Also, it is still a goal on this GEP, not sure I found how this GEP suggests to support it?

It was buy not anymore. I'll remove it

> - Support backends with multiple protocols on the same port (ie. h2c/http1)

Author: Actually, K8s supports TCP/UDP protocol on the same port - so I think this requirement stands.

> - Support backends with multiple protocols on the same port (ie. h2c/http1)

Author: > The reason I am bringing this up is that I am trying to collect real uses cases for converting AppProtocol to a list.

An alternative to a list could be allowing multiple entries like how the mixed protocol KEP suggested it

https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1435-mixed-protocol-lb#proposal

> - Support backends with multiple protocols on the same port (ie. h2c/http1)

Reviewer 8: lets say we support a list, how would it help for your use case?

> - Support backends with multiple protocols on the same port (ie. h2c/http1)

Author: Knative OSS and Cloud Run support the same portable API. It has a restrictions to listen only on one port since we need to capture request metrics to scale and we also enforce request concurrency.

So if there was a TensorFlow user which wanted to bring  that workload to Knative in order to get elastic autoscaling and scale to zero etc. they wouldn't be able to host both GRPC and HTTP using the same instance.

The TensorFlow Serving isn't a random use case - there are competing products doing the same thing - https://kserve.github.io

> - Support backends with multiple protocols on the same port (ie. h2c/http1)

Reviewer 4: We run Istio on CloudRun and do the same - gRPC and http on the same port

- Extensible - additional protocol specific fields should be easy to add in the future

## Non-Goals

> - Backend TLS (covered in [GEP-1897](https://github.com/kubernetes-sigs/gateway-api/issues/1897)

Reviewer 9: ```suggestion
- Backend TLS (covered in [GEP-1897](https://github.com/kubernetes-sigs/gateway-api/issues/1897))
```

that operate at a higher level (eg. HTTP/GRPC etc.)

### Kubernetes Service - Expanding AppProtocol field

Since Kubernetes 1.20 the core/v1.Service object has a stable `appProtocol` field. It's purpose is to allow end-users to
specify an application protocol (L7) for each service port.

Originally the use of this field in the Gateway API was rejected in [GEP-1282](geps/gep-1282.md#non-goals):
> v1.Service’s appProtocol field is not fit for purpose, because it is defined as accepting values either from the IANA Service Name registry, or domain-prefixed values and we need more flexibility than that. 

Further discussion [here](https://github.com/kubernetes-sigs/gateway-api/discussions/1244#discussioncomment-3151197)

Since then a Kubernetes enhancement proposal was created [KEP-3726](https://github.com/kubernetes/enhancements/issues/3726)
to repurpose `appProtocol` to include a convention for protocols that are not IANA service names. This would involve
prefixing protocol names with `kubernetes.io/`. At the moment there exists two defined protocols `kubernetes.io/h2c` and `kubernetes.io/grpc`.

> It is not possible to have different `appProtocol` entries for the same port. A related [KEP-1435](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1435-mixed-protocol-lb) mentions different `protocol` entries for the same port. Finally it's unclear how one would disable certain protocol behaviours (ie. HTTP/1 Upgrades).

Author: I know envoy has options to disable this in various places - https://www.envoyproxy.io/docs/envoy/v1.26.1/intro/arch_overview/http/upgrades.html

I don't have a use case really - just added this clause since `appProtocol` mechanic allows enabling but not disabling.

> It is not possible to have different `appProtocol` entries for the same port. A related [KEP-1435](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1435-mixed-protocol-lb) mentions different `protocol` entries for the same port. Finally it's unclear how one would disable certain protocol behaviours (ie. HTTP/1 Upgrades).

Author: I made this a non-goal for the GEP

Note: Kubernetes will automatically create `EndpointSlices` for `Services` that have a selector. [Custom `EndpointSlices`](https://kubernetes.io/docs/concepts/services-networking/service/#custom-endpointslices) can manually be created.

## API Semantics

A Gateway implementation MUST recognize the Kubernetes Standard Application Protocols ([KEP-3726](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3726-standard-application-protocols)) for specifying the protocol for a backend reference in a Gateway API `*Route` resource

Thus when a `*Route` points to a Kubernetes Service resource the backend protocol for each port can be specified by
- setting the `appProtocol` field on the Kubernetes `Service`
- setting the `appProtocol` field on an `Endpoint`/`EndpointSlice` object associated with a Kubernetes `Service`

At the moment there exists two defined constants:
- `kubernetes.io/h2c` - HTTP/2 over cleartext as described in https://www.rfc-editor.org/rfc/rfc7540

> - `kubernetes.io/raw` - Instructs the implementation to look at the `Service`'s `protocol` field which supports `TCP`, `UDP`, and `SCTP`. Default is `TCP`.

Reviewer 8: `kubernetes.io/raw` - was not recognized  as a standard (yet). I know the current situation is confusing because https://github.com/kubernetes/enhancements/pull/3912 was not merge yet to reflect the actual implementation. 
I will ping again there to ask for approval.

Author: PR merged so going to resolve this

Reviewer 8: can you remove `kubernetes.io/raw`? it is not a standard so it may confuse other people.

Author: What's the alternative mechanic to say we support TCP etc. ?

Reviewer 8: The current situation is the if the protocol is `TCP` and `appProtocol` is not set, implementations do whatever they want, one can take is "raw" TCP other can try to sniff and guess the traffic.

We wanted to introduce an appProtocol value that will basically be "raw" not only for TCP but for all/most port protocols (TCP/UDP/SCTP) hence the name `raw` and not `tcp-raw` or `tcp`.  That said, it was last minute and we did not added it eventually. (https://github.com/kubernetes/kubernetes/pull/115433#discussion_r1136267401)

You could add it to your current open enhancement and/or start a discussion in slack/github to see what other people think.

Author: I added a PR to add support for raw - https://github.com/kubernetes/enhancements/pull/4106

Author: Resolving as raw is not supported upstream - https://github.com/kubernetes/enhancements/pull/4106

Will remove it from the GEP

### Multiple Protocols on the same port

Only the Kubernetes `Service` `protocol` field supports multiple protocols on the same port. See the details in [KEP-1435](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1435-mixed-protocol-lb).

Implementations MUST recognize support for multiple protocols on the same port.

Currently Kubernetes `Service` API does not allow different `appProtocol` values for the same port. At this time there seems to be interest in changing `appProtocol` to be a list in order to faciliate this use-case.

### Default Protocols

If a backend protocol isn't specified an implementation MAY infer the backend protocol through its own means. 

### Unsupported/Incompatible Protocols

> If a Route attached to a Gateway is not able to sending traffic to the backend using the specified protocol then it must set it's RouteConditionType `Accepted` to `False`, with a RouteConditionReason `UnsupportedProtocol`

Reviewer 8: To clarify, `kubernetes.io/raw` is still not supported. (see [here](https://github.com/kubernetes-sigs/gateway-api/pull/1979#discussion_r1214081474))

I do think it is useful to have one but it would need an small enhancement.

Aside from that, I am not sure I understand the examples.  Why would HTTPRoute not get https, http or `kubernetes.io/h2c` ?

> If a Route attached to a Gateway is not able to sending traffic to the backend using the specified protocol then it must set it's RouteConditionType `Accepted` to `False`, with a RouteConditionReason `UnsupportedProtocol`

Author: I created at able showing some recommended compatability - it does raise more questions

### Supporting new protocols

> To add support for a new protocol it should first become a Kubernetes Standard Application Protocol by updating the [KEP-3726](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/3726-standard-application-protocols). If the suggested protocol is not suited to have a `kubernetes.io/*` prefix then the Gateway API MAY support the new protocol using it's own prefix `gateway-api.kubernetes.io/*`.  Please make a PR to this GEP.

Reviewer 7: Can you please split this on multiple lines?

Aside from that, love this, this is a great section to have in a GEP, and we should consider adding it to other GEPs in general - "How to add new bits to this GEP" is not clear in many cases.

## Alternatives

### Single Meta-resource 

The first pass of this GEP proposed a new meta-resource [GEP-713](geps/gep-713.md) called `BackendProtocol`.

This allows end-users to specify a list of ports and a list of corresponding protocols that that single
port supports.

This was dropped in favour of supporting Kubernetes Standard Application Protocols. The Go and YAML API are
presented below for posterity.

#### Go API

```go
type BackendProtocol struct {
  // Standard type metadata
  metav1.TypeMeta `json:",inline"`

  // Standard object's metadata
  metav1.ObjectMeta `json:"metadata,omitempty"`

  // Spec defines the behaviour of this meta-resource
  Spec BackendProtocolSpec `json:"spec,omitempty"`
}

type BackendProtocolSpec struct {
  // Specifies the API resource who's protocols are being
  // explicitly defined
  TargetRef LocalObjectReference `json:"targetRef"`

  // Ports contains a list of protocols for a specific ports
  // When multiple protocols reference the same port the 
  // entries earlier in the list should have higher priority
  //
  // ie. attempt to use h2c otherwise fallback on http1
  //  ports: 
  //   - port: 80
  //     http2: 

>   //       encrypted: false # enable h2c

Reviewer 10: Seems a bit strange - h2c has pretty clear semantics. 

Also differentiating between false and not set can be confusing

Reviewer 10: Also if encrypted = true  - we need to be clear ALPN takes priority and any setting here is ignored.

Author: you're reviewing an alternative proposal - this isn't what's being recommended anymore

  //   - port: 80
  //     http1: {}          # We also listen to http1 on port 80
  //
  Ports []BackendPort `json:"ports,omitempty"`
}

type BackendPort struct {
  // Number or name of the port who's protocol is being defined
  Port intstr.IntOrString

  PortProtocol `json:",inline"`
}

type PortProtocol struct {

>   HTTP2     *HTTP2PortProtocol    `json:"http2"`

Reviewer 10: How about adding a separate 'TLS' protocol - which indicates the backend handles TLS and will negotiate the protocol as defined in the RFC.

Author: you're reviewing an alternative proposal - this isn't what's being recommended anymore

  HTTP1     *HTTP1PortProtocol    `json:"http1"`

>   GRPC      *GRPCPortProtocol     `json:"grpc"`

Reviewer 10: Is it encrypted ? How is it different from h2c from gateway perspective ?

Author: you're reviewing an alternative proposal - this isn't what's being recommended anymore

  WebSocket *WebsocketPortProtocol`json:"websocket"`

> }

Reviewer 10: Not sure what websocket means. 

WS works over h2 - as CONNECT method - and on http1.1, where the protocol upgrade happens. I assume this is the http/1.1 websocket ? But even on 1.1, WS is based on path - a backend may handle both regular POST/GET methods and have WS on one path. 

Same for http2, where CONNECT is just one of the methods that a h2 server supports, so can't be just ws.

There is also the question of what will the gateway do when receiving a h1 websocket request and the backend is h2, or the other way around.

Author: you're reviewing an alternative proposal - this isn't what's being recommended anymore

> v1.Service’s appProtocol field is not fit for purpose, because it is defined as accepting values either from the IANA Service Name registry, or domain-prefixed values and we need more flexibility than that. 

Since then a Kubernetes enhancement proposal was created [KEP-3726][kep-3726] to repurpose `appProtocol` to include a convention for protocols that are not IANA service names. This would involve prefixing protocol names with `kubernetes.io/*`.

Note: Kubernetes will automatically create `EndpointSlices` for `Services` that have a selector. [Custom `EndpointSlices`](https://kubernetes.io/docs/concepts/services-networking/service/#custom-endpointslices) can manually be created.

## API Semantics

A Gateway implementation MUST recognize the Kubernetes Standard Application Protocols ([KEP-3726][kep-3726]) for specifying the protocol for a backend reference in a Gateway API `*Route` resource

Thus when a `*Route` points to a Kubernetes Service resource the backend protocol for each port can be specified by
- setting the `appProtocol` field on the Kubernetes `Service`
- setting the `appProtocol` field on an `Endpoint`/`EndpointSlice` object associated with a Kubernetes `Service`

> At the moment there exists two defined constants:

Reviewer 11: ```suggestion
At the moment there exists three defined constants:
```

### Default Protocols

If a Service `appProtocol` isn't specified an implementation MAY infer the backend protocol through its own means. Implementations MAY infer the protocol from the `Route` type referring to the backend Service.

### Multiple Protocols on the same port

Only the Kubernetes `Service` `protocol` field supports multiple protocols on the same port. See the details in [KEP-1435][kep-1435].

Implementations MUST recognize support for `TCP` and `UDP` multiplexing on the same port.

Currently Kubernetes `Service` API does not allow different `appProtocol` values for the same port. At this time there seems to be interest in changing `appProtocol` to be a list in order to faciliate this use-case.

Open Question: Can SCTP be multiplexed with TCP and UDP on the same port?

### Unsupported/Incompatible Protocols

> If a Route attached to a Gateway is not able to sending traffic to the backend using the specified protocol then it must set its RouteConditionType `Accepted` to `False`, with a `RouteConditionReason` `UnsupportedProtocol`

Reviewer 11: considering that there could be multiple backends in the HTTPRoute, is it reasonable to invalidate the whole resource if there is a problem with just one backend?

Author: Good catch. There already is precendence for invalid backend. I'll follow that.

Author: https://github.com/kubernetes-sigs/gateway-api/blob/d4e08e888f3d19ddbb5242324a6505899fc29e9c/apis/v1beta1/httproute_types.go#L1061-L1064
